teps/0017-results-api.md

```diff
@@ -0,0 +1,711 @@
+periodicly deleting completed resources from their cluster
+(https://github.com/tektoncd/plumbing/issues/439), but this at the cost of
+losing the execution history. Offloading long-term result storage to a separate
+system will enable us to add support to clean up these completed resources to
+help reduce the load on the Pipeline controller, without sacrificing historical
+results.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Pipelines Source of Truth
+
+The Results API is intended for long term result storage for users - it is not
+meant to be a data plane replacement for the Pipeline controller. There should
+be an expectation of eventual consistency - clients that are sensitive to this
+should continue to rely on the Pipeline API directly.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+```
+<<[UNRESOLVED wlynch ]>>
+This section will be filled in based on feedback
+<<[/UNRESOLVED]>>
+```
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+### Proposed Spec
+
+We propose the following API, heavily influenced by API recommendations defined
```

Comment:
I'd like to propose an alternative that's been kicking around in my head, and get your feedback:

Instead, the Results API is a superset of the API-spec-conformant TaskRun/PipelineRun API, with the addition of a non-K8s-standard `filter` field on the API endpoint to list TaskRuns and PipelineRuns (i.e., `GET /apis/tekton.dev/v1beta1/{parent=*}/taskruns`). All other API methods work exactly the same, no changes, except responses are served from mysql instead of etcd.

This makes the K8s API server a slightly degenerate implementation of the Results API, since it serves GetTaskRuns/ListTaskRuns/etc. but just doesn't support `filter`.

The benefit is that taken holistically, a client doesn't have to think about "TaskRuns", "PipelineRuns", and "Results" as separate resources/endpoints, they can just think about the first two and maybe get better indexing/filtering depending on the backend.

This proposal isn't even really at odds with yours, we could probably do both and have this Tekton-API-compliant-plus-`filter` version be a facade for the `Resuls`-focused implementation.

WDYT?

Reply:
Yeah I think you've convinced me that we need Results as a concept, and this is great in going in that direction.

Do you think you want to mention the facade idea in the TEP? Or maybe wait and leave it as a topic for discussion in an issue in the new repo?

Reply:
Let's leave it off for now - it's an additive change and I don't believe it changes what we want to do in this proposal.